cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -302,11 +325,53 @@ private predicate fieldReadStep(Node node1, FieldContent f, Node node2) {
   )
 }
 
+predicate suppressArrayRead(Node node1, ArrayContent a, Node node2) {
+  a = TArrayContent() and
+  exists(BufferMayWriteSideEffectInstruction write, ChiInstruction chi |
+    node1.asInstruction() = write and
+    node2.asInstruction() = chi and
+    chi.getPartial() = write and
+    getWrittenField(write, _, _)
+  )
+}
+
+private class ArrayToPointerConvertInstruction extends ConvertInstruction {
+  ArrayToPointerConvertInstruction() {
+    this.getUnary().getResultType() instanceof ArrayType and
+    this.getResultType() instanceof PointerType
+  }
+}
+
 private predicate arrayReadStep(Node node1, ArrayContent a, Node node2) {
   a = TArrayContent() and
-  exists(LoadInstruction load |
-    node1.asInstruction() = load.getSourceValueOperand().getAnyDef() and
-    load = node2.asInstruction()
+  (
+    // In cases such as:
+    // ```cpp
+    // void f(int* pa) {
+    //   *pa = source();
+    // }
+    // ...
+    // int x;
+    // f(&x);
+    // use(x);
```

Comment:
I'm still not happy to make the `x` in `use(x)` an array read. In my view, the array-like behaviour around `x` should be resolved after the call to `f(&x)` so that there is no confusion in a snippet like
```c
int x;
if (b)
  f(&x);
else
  x = g();
use(x);
```

Reply:
Indeed, I'm also not a fan of this. I'll investigate this for a bit and see if this is something we can easily get rid of.

Reply:
I've removed this case from `arrayReadStep` in 62d42f20d9d16e73e65be9b1c7b2bbd8423f2d8f and replaced it with a slightly less horrible solution (I think).